### ISSUE TITLE
Load and save+publish times increased between 12.0.0 and 12.1.2

### ISSUE THREAD

**Issue by marklonquist** [umbraco/Umbraco-CMS#14842]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

12.1.2

### Bug summary

We have a huge ecom site which uses umbraco for content, and we make heavy use of Block List throughout the site plus have over 20 languages enabled.

On some of our nodes, we see an increase in load-time (GetById) from ~8s to ~60s and save+publish going from ~3m to >10m (times out).

While 3m is also incredible long, it does make sense when you have Block List, within Block List within Block List. Added a video showing the worst node we have.

### Steps to reproduce

Have a site with 20+ languages, all with content, and deep nesting of Block Lists

https://github.com/umbraco/Umbraco-CMS/assets/1456819/bb24f1f5-f965-4cf6-b906-2576004ad12c

### Expected result / actual result

Faster loading and faster saving

**Comment by Zeegaan** [comment]
Hey I'm sorry you're experiencing that 😞 
However I don't really have a giant site with 20+ languages on hand 😁 

I would love if you could recreate this from a clean site, and then perhaps send me the backpack file of your database, that way it would be easy for me to test & reproduce 🤞 

Even a database file of current database file would help 😁

**Comment by marklonquist** [comment]
@Zeegaan I created a repository with version that can load and save+publish the node we're really struggling with.

See the README.md

**Comment by marklonquist** [comment]
Hi @Zeegaan - you had time to play around with this? :p

**Comment by Zeegaan** [comment]
@marklonquist Yep just cloned it down, we need the username + password for the backoffice user though 😁

**Comment by marklonquist** [comment]
Hi @Zeegaan - what email can I send the username/password to?

**Comment by Zeegaan** [comment]
@marklonquist nge@umbraco.dk

**Comment by andr317c** [comment]
Hey! I tried looking at your issue. I was able to reproduce the high load times.
As you explained in the readme, I tried going to Configuration > Navigation Container > Burger Menu in the content section, to see how long it took to load, both for versions 12.0.0 and 12.1.2.

- Version 12.0.0 Load time
  - First try: 28 seconds
  - Second try: 27 seconds
- Version 12.1.2 Load time
  - First try: 4:13 minutes
  - Second try:  4:44 minutes
  
The load times seem to be around 9-10 times slower in version 12.1.2 than it was in 12.0.0.
 
I was not able to reproduce the slow Save and Publish time because I ran into this error 
![image](https://github.com/umbraco/Umbraco-CMS/assets/73799582/8dac26ff-91ae-4ed1-81be-7c5714ea4538)

I will mark this issue as a sprint candidate. 

Please let me know what I am missing so I'm able to Save and Publish! 😄

**Comment by marklonquist** [comment]
@Zeegaan I added the missing property editor - seems to be able to save and publish now

**Comment by andr317c** [comment]
Hey again, I tried running the project again with the missing property editor. I ran into another issue when trying to Save and Publish, this one: 
![image](https://github.com/umbraco/Umbraco-CMS/assets/73799582/5aea142e-44b8-40c8-81de-7bb4d413c939)
Save and Publish was running for 23 minutes before I encountered this error. So I can confirm that the Save and Publish times are a lot longer in 12.2.0 than in 12.0.0 for your project

**Comment by dawoe** [comment]
Hi @marklonquist, @Zeegaan, @andr317c 

Could be related to this issue (#14936 ). We experiencing a slow back office as well when making use of the block list with a lot of culture and segment variants.

**Comment by bielu** [comment]
So had debugging session and figure out a lot of slowiness will come from similar issue which was caused by block editor, which would be solved by pr, a lot of of slowiness will come also from PropertyValidationService which will load datatype for each property in each culture 😱, and based on that what i learned from fixing Block  editor it has a lot of impact on performance, so going down umbraco loads almost 1.3k duplicated datypes in my specific case! so it would add up up to almost 2s!

**Comment by nielslyngsoe** [comment]
Hi @Migaroez 

Does this PR https://github.com/umbraco/Umbraco-CMS/pull/15731 fixes this or is there more that can be done — this issue is getting a bit stale so would be nice to give an update on the matter. Thanks in advance.

**Comment by marklonquist** [comment]
We are on 13.4.0-rc and we still have issues with long save+publish times

**Comment by marklonquist** [comment]
Does seem like it's way better in Umbraco 15. Just saved 86 languages, each having 2000 blocks (so 172000 total blocks) in 45 seconds.. That's pretty good.

**Comment by bielu** [comment]
@marklonquist v15 is not nucache based so same issue is not applying as memory cache is differently used. Also I wouldn't use v15 in production tbh.

**Comment by marklonquist** [comment]
> @marklonquist v15 is not nucache based so same issue is not applying as memory cache is differently used. Also I wouldn't use v15 in production tbh.

we have a major assignment in migrating from v13, so there will be plenty of updates to v15 before we are done :P It should mature enough before we are done :D

**Comment by bielu** [comment]
@marklonquist I would personally wait till 17.3 or 5 but yeah that's not suitable case for most

**Comment by marklonquist** [comment]
After some more testing, Umbraco is still slow with a lot of lovalized content using block list.

Sadly, we still migrating to other CMS because of this. 😔

**Comment by bielu** [comment]
@marklonquist did you try to profile why it is slow? As I am curious as v15 shouldn't share as much this issue as V13 did 🤔

**Comment by marklonquist** [comment]
With ~90 languages and block lists inside block lists, it just seem to be having issues. No profiling. Just a comparison between our current v13.4 instance and a fresh v15.

Could try to profile it.

**Comment by bielu** [comment]
I am not that familar with v15 code but you can try https://github.com/umbraco/Umbraco-CMS/pull/15573 similar steps to mine from v13 :) I had a lot fun learning that memory cache done by umbraco cause delays by over locking memory cache

**Comment by marklonquist** [comment]
Did some profiling. It seems it's just because the amount of SQL calls is big (one per culture)

[GET https___localhost_44357_api_testdata (4602.63 ms) - Profiling Results_compressed.pdf](https://github.com/user-attachments/files/18006626/GET.https___localhost_44357_api_testdata.4602.63.ms.-.Profiling.Results_compressed.pdf)

**Comment by bielu** [comment]
@marklonquist that result seem to be worst than v13 about amount of sql queries 😱, As you already prepared test data can you share it? i might have a look into it as wanted get familiar with v15 code base anyway lol 
Can you also confirm if v13 was faster/slower?

**Comment by marklonquist** [comment]
@bielu the performance between v13 and v15 seems to be the same in terms of this issue :)

**Comment by nielslyngsoe** [comment]
Hi @marklonquist, First, thanks for the update and sharing your insights.

I'm not sure why this one has gone stale; it seems pretty important to me.

Therefor I have assigned someone to look into this specific issue next week, so if you have any additional details you find relevant, please let us know.

**Comment by AndyButland** [comment]
I'll close this one now, mainly as the original issue is raised for particular versions of Umbraco 12 which is out of support now.

We have used this and other reports in recent efforts though to diagnose and improve performance issues, which has yielded improvements via the following PRs:
- https://github.com/umbraco/Umbraco-CMS/pull/20485
- https://github.com/umbraco/Umbraco-CMS/pull/20405
- Those linked from https://github.com/umbraco/Umbraco-CMS/issues/18892

We'll of course take more specific issue reports and address further where we can moving forward.
